src/java-opentimelineio/src/main/java/io/opentimeline/opentimelineio/SerializableObject.java

```diff
@@ -0,0 +1,83 @@
+package io.opentimeline.opentimelineio;
+
+import io.opentimeline.OTIONative;
+
+public class SerializableObject extends OTIONative {
+
+    public SerializableObject() {
+        this.initObject();
+    }
+
+    public SerializableObject(long nativeHandle) {
+        this.nativeHandle = nativeHandle;
+    }
+
+    private void initObject() {
+        this.className = this.getClass().getCanonicalName();
+        if (this.className.equals("io.opentimeline.opentimelineio.SerializableObject"))
+            this.initialize();
+    }
+
+    private native void initialize();
+
+    public native boolean toJSONFile(String fileName, ErrorStatus errorStatus);
+
+    public native boolean toJSONFile(String fileName, ErrorStatus errorStatus, int indent);
+
+    public native String toJSONString(ErrorStatus errorStatus);
+
+    public native String toJSONString(ErrorStatus errorStatus, int indent);
+
+    public static native SerializableObject fromJSONFile(String fileName, ErrorStatus errorStatus);
+
+    public static native SerializableObject fromJSONString(String input, ErrorStatus errorStatus);
+
+    public native boolean isEquivalentTo(SerializableObject serializableObject);
+
+    public native SerializableObject clone(ErrorStatus errorStatus);
+
+    public native AnyDictionary dynamicFields();
+
+    public native boolean isUnknownSchema();
+
+    public native String schemaName();
+
+    public native int schemaVersion();
+
+    public static class Retainer<T extends SerializableObject> extends OTIONative {
+
+        public Retainer(long nativeHandle) {
+            this.nativeHandle = nativeHandle;
+        }
+
+        public Retainer(T value) {
+            this.initialize(value);
+        }
+
+        public Retainer(Retainer<T> retainer) {
+            this.initialize(retainer.value());
+        }
+
+        private native void initialize(T value);
+
+        public native T value();
+
+        public native T takeValue();
+
+        private native void dispose();
+
+        @Override
+        public void close() throws Exception {
+            dispose();
+        }
+    }
+
+    public native int currentRefCount();
+
+    private native void possiblyDispose();
```

Comment:
possiblyDispose or just dispose? Is there a difference for the two?

Reply:
Yes. SerializableObject does not have a public destructor in C++. We cannot delete a `SerializableObject*` or anything derived from it  directly. Instead we need to call the `possibly_delete` function.

This `possibly_delete` function checks before deleting if there  is anything holding a reference to it. And if possible it'll deletes recursively all other objects owned by this object. Like if there's a track with a clip, and the track is the sole owner of the clip, deleting the track should delete the clip too, and if the clip is added to multiple tracks, calling the `possibly_delete` function will decrease its reference count by one and not actually delete the object.

Is this correct @reinecke  or am I missing something?

Reply:
So SerializableObject is a container and if it holds an object that is referred by other containers or in code, it will not dispose those, did I understand this correctly?

For the library user, I would still call this `dispose` method, just flag with a comment that it will not clean up references held by other classes. In general all of our OTIONative objects should be going through Factory and will be natively GC-ed and disposed, so this would just be a dispose implementation of a "collection" class. Does this make sense?

Reply:
f you haven’t read the rationale behind the C++ memory scheme, and in particular, how it has to interoperate with Python, I strongly suggest reading it before going further.

Quite a lot of the complication is having two sets of objects: a C++ underlying object and then a “shadow” or “wrapper” in the language you’re working with.

I believe that a lot of the issues in wrapping in Python may be similar to wrapping in Java may be similar to wrapping in Swift.

In particular, Python/Swift are automatically reference counted, where C++ is not, so we had to develop techniques that worked well together.  If Java is inherently reference counted, maybe some of the same lessons would help you out.

Reply:
I have gone throught the explanation here: https://opentimelineio.readthedocs.io/en/stable/cxx-proposal/cxx.html#memory-management

From what I understand, I need to keep a cpp Retainer reference for the Java object I create. And call `possibly_delete` whenever the java object is deleted. But what I don't understand is what is to prevent something on the C++ side to delete/decrease the reference count while the java/python object still exists? Also, is there a case where something like this could happen?

Reply:
The point of keeping that extra C++ Retainer object, created and owned specifically and only by the Java wrapper is that that extra retainer ensures the C++ object cannot be deleted.  When the Java objects goes away, it destroys that C++ Retainer object.  You have to write some small extra C++ code obviously that dynamically allocates a C++ Retainer, and then (I assume) hands back a pointer to it to be held by Java.  When the Java object goes away, I assume it calls some more custom C++ code you write that takes the pointer and causes the C++ object to be destroyed (and thus the destructor of the C++ Retainer object to fire).  It is the cleanup of that C++ retainer that then allows the C++ SerializableObject to finally die.

Reply:
Here is a longer discussion about how the Python/C++ interoperates.  I could have sworn this discussion had gone someplace more central, but I can't find it.  This is preserved in my own private cut of a readthedocs bundle:

https://opentimelineio-deb.readthedocs.io/en/latest/cxx-proposal/cxx.html#extended-memory-management-discussion

You're looking for "Extended Memory Management" discussion.

Reply:
Yes, this is exact what I've done to create and delete the objects. The `possiblyDispose` you can see in the above snippet is that extra pice of cpp code that causes the cpp object to be destroyed.

Reply:
My bad; my brain saw possiblyDispose() and thought you meant C++ possiblyDelete().  Great.

Reply:
Oh, you were right. I didn't receive the Extended discussion comment when I said "this is what I've done to delete the objects.". I said that based on your previous comment. I still need to implement the keep alive object.

Reply:
I am going through the implementation and I think I am getting a hang of what's going on. 
One thing I'm unable to find: When the python object gets deleted, is there a destructor that calls the possibly_delete function to delete the c++ object?

It's taking some time understanding how the reference counting mechanism works, now that I came to know about the keepalive monitor.

Reply:
If I have a clip and a track. I add the clip to the track.

Later I use a getter method to get the children from the track, does it return a new python object that holds a handle to that clip?

Reply:
Creation of a Python object P1 that mirrors a C++ OTIO object C1 creates a Retainer<> that stays alive as long as P1 stays alive.  When P1 goes away, the Retainer<> goes away, and the destruction of the Retainer<> does the same thing that calling possiblyDelete() does.  So you won't find an explicit code path related to the Python that is calling possiblyDelete(); it is implicit in destruction of the Python object P1 which ends up destroying the Retainer<> associated with P1.

As to your second question about clip/track:   The very first time a C++ OTIO object C1 is handed back to Python, we create a new Python object P1.  For as long as C1 exists, repeated attempts to get a Python wrapper for C1 will always return the exact same P1.

Note: we found this necessary in Python for 2 reasons.  (1) Because people might write new OTIO classes in Python, we needed this to be so, even though we're not really sure this is going to happen.  (2) A stronger reason is that if you stick dynamic attributes on P1, you'd like them to persist and not vanish if you happen to lose P1, but then later recreate on get a new handle to it in Python.

In Java, if you're not worried about extending functionality by writing new classes in Java (which is probably unlikely, but just my opinion) AND if being able to just add dynamic data to your Java instance is not a concern, you might relax this restriction.  Said another way, if multiple requests to get the a Java object from C1 return different instances J1 and J2 and there is never any functional difference between them, you might want to drop this restriction.  If you do, it makes the implementation much simpler.

Reply:
I don't think we can add dynamic data to a Java object the way it can be done in python.

How is it that the same python object is returned? Is some record maintained somewhere or is it handled by pybind11?

Reply:
It's handled by pybind11; as long as the Python wrapper object P1 pybind11 handed back has not been destroyed, pybind11 will hand it back again.  The trick is that it's up to us to keep P1 alive for as long as it *might* need to be returned back to Python, while allowing it to be destroyed when it is clear that it *cannot* be handed back again.

Reply:
There might be one issue with handing out multiple java objects that hold a reference to the same C++ object.

If we follow the new design pattern that @margant suggested. Instead of each Java class extending OTIONative (would be analogous to `struct managing_ptr` in C++), each class could hold an OTIONative object. 

If we have C++ object C1, and corresponding Java object J1, then we get J2 which is functionally same as J1 and holds reference to C1. In case J2 is deleted, C1 would get deleted because J2 was just a copy of J1 and when fetching J2 the reference count didn't increase. Later, if J1 was to get deleted, it would again try to delete the already deleted C1. This would be wrong.

Does this make sense?

Reply:
Yes, but this is just a standard value type vs reference type semantics issue.  If a Java instance J1 holds an external resource, and J1 is a reference type (meaning that you don't copy it but just bump it's reference count) then everything is fine.  If however the type has value semantics, then you can't just blindly copy its contents: you need to copy in such a way that you get a new distinct C++ Reference<> object.

Same applies if your type was holding, say, a mutex.  This is just a basic issue hardly unique to OTIO concerns.

Reply:
> It's handled by pybind11; as long as the Python wrapper object P1 pybind11 handed back has not been destroyed, pybind11 will hand it back again. The trick is that it's up to us to keep P1 alive for as long as it might need to be returned back to Python, while allowing it to be destroyed when it is clear that it cannot be handed back again.


```
void monitor() {
        pybind11::gil_scoped_acquire acquire;
        if (_so->current_ref_count() > 1) {
            if (!_keep_alive) {
                _keep_alive = pybind11::cast(_so);
            }
        }
        else {
            if (_keep_alive) {
                _keep_alive = pybind11::object();      // this could cause destruction
            }
        }
    }
```

So if we have P1 and we create this keepalive monitor, by doing `pybind11::cast(_so)`, `_keep_alive` would be the exact same object?

If so, it's not the same with Java. JNI is just a Java-Native interface. It doesn't do any such management on its own.

Reply:
It might be possible to keep track of memory address of ("this") in Java within the OTIONative object, the trick would be to keep that up to date as it gets GC-ed. The only way to dispose of java object is by GC, there is no way to do that from C++, but if you register all Java objects with OTIONative references from the factory, they should all be cycled through the cleanUp, so we should be able to get to the native call to remove that memory address (aka ref count as above). 

Having said that, there is a huge design movement for immutable objects so if we could enable 1-1 java to OTIONative instance and not allow that to be shared, I think that should be still fairly solid from the design perspective. Personally I would even prefer not sharing the C++ object around across multiple instances (aka always deep copy) since mutations are notoriously difficult to keep track when done from multiple angles.